Add model_size_bytes attribute when saving the model

wenfeiy-db opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/wenfeiy-db/mlflow/pull/10110?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10110/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10110
```

</p>
</details>

### What changes are proposed in this pull request?

Add `model_size_bytes` attribute in the mlflow model (thus also in `MLmodel` file), which is calculated as the size of everything saved in the `local_path` (later logged as artifacts) inside `save_model`. 

The attribute is added to all flavors with a few exceptions below
* OpenAI: IMO we should not add model_size_bytes for this flavor, as it’s a remote model and we don’t know the actual model size
* Deprecated flavors - MLleap and MXNet Gluon
* R Function: My understanding is that it'll need a different implementation in R, and I'd like to leave it out of the scope of this PR.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

wenfeiy-db committed:
transformers

Signed-off-by: wenfeiy-db <wenfei.yan@databricks.com>

wenfeiy-db committed:
pyfunc

Signed-off-by: wenfeiy-db <wenfei.yan@databricks.com>

wenfeiy-db commented on mlflow/models/model.py:
```diff
@@ -598,6 +598,8 @@ def log(
             ):
                 _logger.warning(_LOG_MODEL_MISSING_SIGNATURE_WARNING)
             flavor.save_model(path=local_path, mlflow_model=mlflow_model, **kwargs)
+            # TODO: can we move saving MLModel file here? This way we can set model_size_bytes
+            #  here for all flavors here, instead of having to implement it in each save_model
```
@dbczumar Would love your inputs here!

BenWilson2 replied:
The proposal is to call this method on the `mlflow_model` `Model()` instance within the `save_model()` function in each flavor instead of calling `save()`? 

This method does additional things compared to `save()`. 

I think for abstraction purposes, having an addition to Model().save(), calling a utility function that calculates the file sizes within the target directory that the MLModel is being saved to would perhaps be a bit more appropriate to maintain consistency between model logging and model saving (since logging a flavor will call the save logic within a flavor).

wenfeiy-db replied:
(Synced offline earlier) We agreed on it's actually better to keep calling the utility function from each flavor's own `save_model` for better flexibility and extensibility.

wenfeiy-db committed:
precommit

Signed-off-by: wenfeiy-db <wenfei.yan@databricks.com>

BenWilson2 commented on mlflow/utils/file_utils.py:
```diff
@@ -999,3 +1001,22 @@ def chdir(path: str) -> None:
+        )
+
+    total_size = 0
+    for cur_path, dirs, files in os.walk(path):
```
Does this work for all file types? 

I tried this logic on my local MLflow filesystem storage. 

```python 
def get_total_size(path: str) -> int:
    """
    Return the size of all files under given path, including files in subdirectories.
    :param path: The absolute path of a local directory.
    :return: size in bytes.
    """
    if not os.path.isdir(path):
        raise MlflowException(
            message=f"The given {path} is not a directory.", error_code=INVALID_PARAMETER_VALUE
        )

    total_size = 0
    for cur_path, dirs, files in os.walk(path):
        full_paths = [os.path.join(cur_path, file) for file in files]
        total_size += sum([os.path.getsize(file) for file in full_paths])
    return total_size
```
(I have some large LLM's in here for local OSS testing)
``` python
get_total_size(os.path.expanduser("~/JupyterNotebooks"))
```

```bash
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
/var/folders/cd/n8n0rm2x53l_s0xv_j_xklb00000gp/T/ipykernel_37720/1521265667.py in <cell line: 1>()
----> 1 get_total_size(os.path.expanduser("~/JupyterNotebooks"))

/var/folders/cd/n8n0rm2x53l_s0xv_j_xklb00000gp/T/ipykernel_37720/3741259608.py in get_total_size(path)
     13     for cur_path, dirs, files in os.walk(path):
     14         full_paths = [os.path.join(cur_path, file) for file in files]
---> 15         total_size += sum([os.path.getsize(file) for file in full_paths])
     16     return total_size

/var/folders/cd/n8n0rm2x53l_s0xv_j_xklb00000gp/T/ipykernel_37720/3741259608.py in <listcomp>(.0)
     13     for cur_path, dirs, files in os.walk(path):
     14         full_paths = [os.path.join(cur_path, file) for file in files]
---> 15         total_size += sum([os.path.getsize(file) for file in full_paths])
     16     return total_size

~/miniconda3/envs/mlflow-dev-env/lib/python3.8/genericpath.py in getsize(filename)
     48 def getsize(filename):
     49     """Return the size of a file, reported by os.stat()."""
---> 50     return os.stat(filename).st_size
     51 
     52 

FileNotFoundError: [Errno 2] No such file or directory: '/Users/benjamin.wilson/JupyterNotebooks/mpt-7b/pytorch_model-00001-of-00002.bin'
```

Modifying the approach to use unix-native processing: 

```python
import subprocess
import os

def get_directory_size(path: str) -> int:
    if os.name == 'posix':  # Unix/Linux/macOS
        cmd = ['du', '-sk', path]
        output = subprocess.check_output(cmd)
        size = int(output.split()[0])
    elif os.name == 'nt':  # Windows
        cmd = ['robocopy', '/L', '/XJ', '/E', '/BYTES', '/NP', path, 'NULL']
        output = subprocess.check_output(cmd, universal_newlines=True)
        size_line = [line for line in output.splitlines() if "Bytes :" in line][0]
        size = int(size_line.split(":")[1].strip().replace(",", ""))
    else:
        raise ValueError("Unsupported OS")
    return size
```

``` python
get_directory_size(os.path.expanduser("~/JupyterNotebooks")) / 1000
```

```bash
27766.1
```

wenfeiy-db replied:
Hmm weird - I tested on my local os and it worked
```
get_total_size(os.path.expanduser("~/transformers-cache/models--t5-large-resharded"))
```

I didn't choose  the option of using unix native commands because I'm not super clear what it's returning. It seems to return different director sizes when I do `-s` (which I thought returns bytes) v.s. `-sk` (which I thought returns kilobytes). See below...  
```
(mlflow-dev) wenfei.yan universe [master]$ du -s /Users/wenfei.yan/transformers-cache/models--t5-large-resharded
6277552	/Users/wenfei.yan/transformers-cache/models--t5-large-resharded
(mlflow-dev) wenfei.yan universe [master]$ du -sk /Users/wenfei.yan/transformers-cache/models--t5-large-resharded
3138776	/Users/wenfei.yan/transformers-cache/models--t5-large-resharded
```

BenWilson2 replied:
`-s` is for "summarize" (adds up the individual files sizes). 
-k would be kB, -m for MB, etc. 

Without putting a value in, it uses the default block size for the `du` operator (512bytes if POSIXLY_CORRECT is set as a sys env var, 1024 otherwise). 

I'm not recommending to use this; more very curious about why the implementation wasn't able to read a .bin file in my local filesystem for transformers models. We should definitely investigate thoroughly for correctness to ensure that size operations can work (and perform error handling if there are permissions errors or unreadable files for a given operating system)

wenfeiy-db replied:
Yeah I'm also curious why it'd fail to read the .bin file in your case. Could you describe your setup? I'll try to reproduce.

It's actually a good call that we should try catch the error because failing to calculate the size should not fail the whole `save_model`.

BenWilson2 replied:
The model that failed to read was read using: 
``` bash
mlflow==2.7.1 
transformers==4.34.1
huggingface_hub==0.17.3
``` 

Python 3.8.10 (our min required dev environment for MLflow for compatibility, so that's where the `os` version comes from) 

OSX Ventura 13.5 
M1 Mac Pro


On the try catch front, what is your suggested approach if we get an incomplete read on the directory size? Do we omit the MLModel entry or provide an indicator that the file size could not be calculated properly? 
Is there a hard dependency that you're building on the consumption of the presence of this key downstream? Does typing matter for this field?

wenfeiy-db replied:
My proposal is just omitting this entry in MLModel and optionally log a warning (maybe it's not even worth the "WARNING" level).
The downstream task should not have hard dependency. Anyway models logged from older mlflow versions won't have this model size in MLModel file.

Regarding for different file types and different platforms, my impression is that I can write some more test cases that cover different flavors (thus different file types) and CI will run it on various envs/platforms, and so the problem is covered. Is this correct?

harupy replied:
@BenWilson2 

Does `FileNotFoundError: [Errno 2] No such file or directory: '/Users/benjamin.wilson/JupyterNotebooks/mpt-7b/pytorch_model-00001-of-00002.bin'` exist?

BenWilson2 replied:
@harupy the file exists and I was able to load the saved transformers library with mlflow and with the transformers library directly by passing in the file path. du inspection works as well. I suspect that there are certain conditions where the python os operations don't always correctly inspect files on disk (this isn't the first time I've found this in my time using Python either).

wenfeiy-db commented on tests/utils/test_file_utils.py:
```diff
@@ -364,3 +365,22 @@ def test_shutil_copytree_without_file_permissions(tmp_path):
+    subdir = os.path.join(root, "subdir")
+    os.mkdir(subdir)
+    files = {
+        "file1.txt": "hello world",  # 11 bytes
```
Ah yeah - this is also something I'd like to have your advice. Is it ok to save some example artifacts in the repo for testing purpose? For `transformers` it can be a hundred of MB. Is there a better practice we follow in mfllow?

BenWilson2 replied:
We definitely don't want to save anything like that in the repo. We always fetch such files from huggingface_hub if we need to validate. You can see examples of us downloading and using transformers models within the transformers unit test suite in MLflow (here's an example: https://github.com/mlflow/mlflow/blob/49415f37edc02da7c8e9ccee2b78e294fe5aa26e/tests/transformers/test_transformers_model_export.py#L980-L989 )

We should definitely not try to store such a large file(s) in the actual repo, though.

wenfeiy-db replied:
Ahh yeah. I'll some more tests cases following this. I was also hesitated to introduce network calls, but I guess that's fine if we're already doing it.

wenfeiy-db committed:
add more tests

Signed-off-by: wenfeiy-db <wenfei.yan@databricks.com>

wenfeiy-db committed:
add try catch and sklearn

Signed-off-by: wenfeiy-db <wenfei.yan@databricks.com>

wenfeiy-db committed:
more flavors

Signed-off-by: wenfeiy-db <wenfei.yan@databricks.com>

wenfeiy-db commented:
@BenWilson2 PTAL. I added some test cases for `.bin` and `.pkl` files - let me now if there are other common types from our native flavors that we'd like to cover.

I also added try catch to avoid the failure of `get_total_size` to fail the entire `save_model` call. I'm adding the logic in each flavor's `save_model` for easier logging and readability instead of inside the get_total_size function. Let me know if it doesn't make sense to you.

wenfeiy-db committed:
Merge branch 'master' of github.com:mlflow/mlflow into artifact-size-metadata

wenfeiy-db committed:
fix?

Signed-off-by: wenfeiy-db <wenfei.yan@databricks.com>

wenfeiy-db committed:
precommit

Signed-off-by: wenfeiy-db <wenfei.yan@databricks.com>

harupy commented on tests/utils/test_file_utils.py:
```diff
@@ -364,3 +370,102 @@ def test_shutil_copytree_without_file_permissions(tmp_path):
     assert dst_dir.joinpath("top-level-file.txt").read_text() == "hi"
+
+
+def test_get_total_size_basic(tmp_path):
+    root = str(tmp_path)
+
+    subdir = os.path.join(root, "subdir")
+    os.mkdir(subdir)
+    files = {
+        "file1.txt": b"hello world",  # 11 bytes
+        "file2.txt": b"This is mlflow testing.",  # 23 bytes
+    }
+    for name, content in files.items():
+        with open(os.path.join(root, name), "wb") as fp:
+            fp.write(content)
+    with open(os.path.join(subdir, "file3.txt"), "wb") as fp:
+        fp.write(b"One file under subdir.")  # 22 bytes
+
+    assert get_total_file_size(root) == 56
+    assert get_total_file_size(subdir) == 22
```
/ai write a function to generate a file with a specified size.

harupy replied:
@wenfeiy-db Can we use this function?

harupy commented on mlflow/utils/file_utils.py:
```diff
@@ -967,3 +969,26 @@ def chdir(path: str) -> None:
         os.chdir(cwd)
+
+
+def get_total_file_size(path: str) -> int:
```
Can we use https://docs.python.org/3/library/shutil.html#shutil.disk_usage?

wenfeiy-db replied:
I don't think `shutil.disk_usage` returns what we want. Given a folder that contain roughly ~99MB files, it returns `usage(total=1000240963584, used=902191882240, free=98049081344)`, which seems to be way off. Actually I don't quite understand what it intends to return here.

harupy replied:
You're right! I misunderstood what `shutil.disk_usage` does :)

harupy commented on tests/utils/test_file_utils.py:
```diff
@@ -364,3 +370,102 @@ def test_shutil_copytree_without_file_permissions(tmp_path):
+        get_total_file_size(path_file)
+
+
+@pytest.fixture
+@flaky()
+def small_qa_pipeline():
+    # The return type of this model's language head is a Dict[str, Any]
+    architecture = "csarron/mobilebert-uncased-squad-v2"
+    tokenizer = transformers.AutoTokenizer.from_pretrained(architecture, low_cpu_mem_usage=True)
+    model = transformers.MobileBertForQuestionAnswering.from_pretrained(
+        architecture, low_cpu_mem_usage=True
+    )
+    return transformers.pipeline(task="question-answering", model=model, tokenizer=tokenizer)
+
+
+def test_get_total_size_transformers(small_qa_pipeline, tmp_path):
+    model_dir = tmp_path.joinpath("model")
+    small_qa_pipeline.model.save_pretrained(save_directory=model_dir)
+    tokenizer_dir = tmp_path.joinpath("components").joinpath("tokenizer")
+    small_qa_pipeline.tokenizer.save_pretrained(tokenizer_dir)
+
+    expected_size = 0
+    for folder in [model_dir, tokenizer_dir]:
+        folder = str(folder)
+        expected_size += _calcualte_expected_size(folder)
+
+    assert get_total_file_size(str(tmp_path)) == expected_size
```
Can we test this in `tests/transformers`?

wenfeiy-db replied:
This is essentially a test for `get_total_file_size` but not transformers. What's the benefit of moving it there

harupy replied:
btw why do we need to test `get_total_file_size` with transformers?

wenfeiy-db replied:
To test it with many different file types that our native flavors use as @BenWilson2 suggested in above [comment](https://github.com/mlflow/mlflow/pull/10110#discussion_r1370745458).

harupy replied:
Can we log a model and read `model_size_bytes` in `tests/transformers`?

BenWilson2 replied:
+1 to what @harupy is suggesting. We should put this integration-style test in transformers for the benefit of CI setup having dependencies within that suite. There are auto-use fixtures that clean the huggingface cache after downloading from the hub to prevent disk issues on the CI runners as well. We don't want to have to copy all of that logic for this otherwise very fast CI suite here.

harupy commented on tests/utils/test_file_utils.py:
```diff
@@ -364,3 +370,102 @@ def test_shutil_copytree_without_file_permissions(tmp_path):
+    return expected_size
+
+
+def test_get_total_size_sklearn(tmp_path):
+    iris = datasets.load_iris()
+    X = iris.data
+    y = iris.target
+    X_df = pd.DataFrame(X, columns=iris.feature_names)
+    X_df = X_df.iloc[:, :2]  # we only take the first two features.
+    y_series = pd.Series(y)
+
+    linear_lr = glm.LogisticRegression()
+    linear_lr.fit(X_df, y_series)
+
+    path = str(tmp_path)
+    pickle_dir = os.path.join(path, "pickle_model")
+    os.mkdir(pickle_dir)
+    with open(os.path.join(pickle_dir, "model.pkl"), "wb") as out:
+        pickle.dump(linear_lr, out, protocol=pickle.DEFAULT_PROTOCOL)
+    expected_size = _calcualte_expected_size(pickle_dir)
+    assert get_total_file_size(pickle_dir) == expected_size
+
+    cloudpickle_dir = os.path.join(path, "cloudpickle_model")
+    os.mkdir(cloudpickle_dir)
+    with open(os.path.join(cloudpickle_dir, "model.pkl"), "wb") as out:
+        cloudpickle.dump(linear_lr, out, protocol=pickle.DEFAULT_PROTOCOL)
+    expected_size = _calcualte_expected_size(cloudpickle_dir)
+    assert get_total_file_size(cloudpickle_dir) == expected_size
```
same comment https://github.com/mlflow/mlflow/pull/10110/files#r1372594109

harupy commented on mlflow/pyfunc/__init__.py:
```diff
@@ -2179,6 +2180,10 @@ def _save_model_with_loader_module_and_data_path(
         model_config=model_config,
     )
+    try:
+        mlflow_model.model_size_bytes = get_total_file_size(str(path))
```
Can we update the `Model` class to add `model_size_bytes`?

harupy replied:
Adding an attribute after object construction like this is error-prone and can lead to AttributeError easily.

wenfeiy-db replied:
pinged you in the offline discussion thread

harupy commented on mlflow/utils/file_utils.py:
```diff
@@ -967,3 +969,26 @@ def chdir(path: str) -> None:
+            message=f"The given {path} is not a directory.", error_code=INVALID_PARAMETER_VALUE
+        )
+
+    total_size = 0
+    for cur_path, dirs, files in os.walk(path):
+        full_paths = [os.path.join(cur_path, file) for file in files]
+        total_size += sum([os.path.getsize(file) for file in full_paths])
+    return total_size
```
/ai make this a one-liner using pathlib.

harupy replied:
/ai Nice. Does `rglob` only yield a file?

BenWilson2 replied:
+1 to rglob with wildcard. Pathlib's traversal algorithm is far more sophisticated and performant than os.walk(). 
@wenfeiy-db can we implement this recommended change? :)

wenfeiy-db replied:
Ahh I assumed the traversal will be similar - let me try changing it to use pathlib

wenfeiy-db replied:
@BenWilson2 actually, can you explain how pathlib's traversal is more performant? It seems `pathlib` operations are normally slower than `os` operations ([source1](https://switowski.com/blog/pathlib/), [source2](https://treyhunner.com/2019/01/no-really-pathlib-is-great/)).

I can let `get_total_file_size` accept both str and pathlib.Path, but I'd like to understand what're other benefits (in additional to it's an one-liner) of using pathlib's rglob instead of os.walk?

harupy replied:
I don't think pathlib is faster. I just wanted to make this change to simplify the code.

BenWilson2 commented on mlflow/langchain/__init__.py:
```diff
@@ -295,6 +295,10 @@ def load_retriever(persist_directory):
+    try:
+        mlflow_model.model_size_bytes = get_total_file_size(str(path))
+    except Exception as e:
+        logger.info(f"Fail to get the total size of {path!s} because of error :{e}")
```
```suggestion
        logger.info(f"Failed to get the total size of {path!s} because of error :{e}")
```
nit

BenWilson2 replied:
(to be applied in all instances)

BenWilson2 commented on tests/helper_functions.py:
```diff
@@ -635,3 +636,28 @@ def clear_hub_cache():
         pass
+
+
+def flaky(max_tries=3):
```
Good idea refactoring this utility test function to the helper_functions file. +1

BenWilson2 commented:
It's looking good! Once the remaining items are addressed, we should be good to go!

wenfeiy-db committed:
move

Signed-off-by: wenfeiy-db <wenfei.yan@databricks.com>

wenfeiy-db committed:
more flavors

Signed-off-by: wenfeiy-db <wenfei.yan@databricks.com>

wenfeiy-db committed:
change message

Signed-off-by: wenfeiy-db <wenfei.yan@databricks.com>

wenfeiy-db committed:
clean

Signed-off-by: wenfeiy-db <wenfei.yan@databricks.com>

wenfeiy-db reviewed (commented):
@BenWilson2 PTAL! I just addressed previous comments, and added the `model_size_bytes` attributes to all flavors with a few exceptions
* OpenAI: IMO we should not add model_size_bytes for this flavor, as it’s a remote model and we don’t know the actual model size
* Deprecated flavors - MLleap and MXNet Gluon
* R Function: My understanding is that it'll need a different implementation in R, and I'd like to leave it out of the scope of this PR.

I also have some hesitation for adding `model_size_bytes` for MLlib models (Spark MLlib & some John Show Labs) and left an in-line comment. My understanding is that calculating the size wouldn't be a big overhead comparing to the time of actually saving the MLlib models, but I'm not 100% sure. We can keep it out of the scope of this PR too if there's any concern.

BenWilson2 commented on mlflow/spark/__init__.py:
```diff
@@ -548,6 +553,10 @@ def _save_model_metadata(
         code=code_dir_subpath,
     )
+    try:
+        mlflow_model.model_size_bytes = get_total_file_size(dst_dir)
```
I don't foresee any issues with SparkML / mllib models size inspection and recording. It won't be a completely apples-to-apples comparison for the purposes of RAM estimates due to compression, but I don't think it hurts anything.

BenWilson2 reviewed (approved):
LGTM! Thanks for the useful addition @wenfeiy-db :D

harupy commented on tests/utils/test_file_utils.py:
```diff
@@ -364,3 +365,35 @@ def test_shutil_copytree_without_file_permissions(tmp_path):
     assert dst_dir.joinpath("top-level-file.txt").read_text() == "hi"
+
+
+def test_get_total_size_basic(tmp_path):
+    root = str(tmp_path)
+    subdir = os.path.join(root, "subdir")
+    os.mkdir(subdir)
+
+    def generate_file(path, size_in_bytes):
+        with open(path, "wb") as fp:
+            fp.write(b"\0" * size_in_bytes)
+
+    file_size_map = {"file1.txt": 11, "file2.txt": 23}
+    for name, size in file_size_map.items():
+        path = os.path.join(root, name)
+        generate_file(path, size)
+    generate_file(os.path.join(subdir, "file3.txt"), 22)
+    assert get_total_file_size(root) == 56
+    assert get_total_file_size(subdir) == 22
+
+    path_not_exists = os.path.join(root, "does_not_exist")
+    with pytest.raises(
+        MlflowException,
+        match="does not exist.",
+    ):
+        get_total_file_size(path_not_exists)
+
+    path_file = os.path.join(root, "file1.txt")
+    with pytest.raises(
+        MlflowException,
+        match="is not a directory.",
+    ):
+        get_total_file_size(path_file)
```
/ai rewrite this using pathlib.Path methods. `tmp_path` is a pathlib.Path object.

harupy commented on tests/transformers/test_transformers_model_export.py:
```diff
@@ -3865,3 +3865,40 @@ def test_basic_model_with_accelerate_homogeneous_mapping_works(tmp_path):
     assert loaded(text) == pipeline(text)
+
+
+def test_qa_model_model_size_bytes(small_qa_pipeline, tmp_path):
+    def _calculate_expected_size(path_or_dir):
+        # this helper function does not consider subdirectories
+        expected_size = 0
+        if os.path.isdir(path_or_dir):
+            for path in os.listdir(path_or_dir):
+                path = os.path.join(path_or_dir, path)
+                if not os.path.isfile(path):
+                    continue
+                with open(path, "rb") as fp:
+                    expected_size += len(fp.read())
+        elif os.path.isfile(path_or_dir):
+            with open(path_or_dir, "rb") as fp:
+                expected_size = len(fp.read())
+        return expected_size
+
+    mlflow.transformers.save_model(
+        transformers_model=small_qa_pipeline,
+        path=tmp_path,
+    )
+
+    # expected size only counts for files saved before the MLmodel file is saved
+    model_dir = tmp_path.joinpath("model")
+    tokenizer_dir = tmp_path.joinpath("components").joinpath("tokenizer")
+    expected_size = 0
+    for folder in [model_dir, tokenizer_dir]:
+        folder = str(folder)
+        expected_size += _calculate_expected_size(folder)
+    other_files = ["model_card.md", "model_card_data.yaml"]
+    for file in other_files:
+        path = str(tmp_path.joinpath(file))
+        expected_size += _calculate_expected_size(path)
+
+    mlmodel = yaml.safe_load(tmp_path.joinpath("MLmodel").read_bytes())
+    assert mlmodel["model_size_bytes"] == expected_size
```
/ai rewrite this by leveraging pathlib.Path methods. Note tmp_path is a pathlib.Path object.

harupy commented on mlflow/utils/file_utils.py:
```diff
@@ -967,3 +969,26 @@ def chdir(path: str) -> None:
+    :param path: The absolute path of a local directory.
+    :return: size in bytes.
+    """
+    if not os.path.exists(path):
+        raise MlflowException(
+            message=f"The given {path} does not exist.", error_code=INVALID_PARAMETER_VALUE
+        )
+    if not os.path.isdir(path):
+        raise MlflowException(
+            message=f"The given {path} is not a directory.", error_code=INVALID_PARAMETER_VALUE
+        )
```
Do we need these checks? Doesn't `os.walk` throw?

wenfeiy-db replied:
No, i remember os.walk just return empty iterator when given path is not a dir.

harupy commented on mlflow/utils/file_utils.py:
```diff
@@ -967,3 +969,26 @@ def chdir(path: str) -> None:
+        )
+
+    total_size = 0
+    for cur_path, dirs, files in os.walk(path):
```
Have you tried https://github.com/mlflow/mlflow/pull/10110/files#r1372662869? It didn't work?

harupy commented on mlflow/diviner/__init__.py:
```diff
@@ -181,7 +185,10 @@ def save_model(
     )
-
+    try:
+        mlflow_model.model_size_bytes = get_total_file_size(str(path))
```
Can we add a setter for `model_size_bytes` in the `Model` class?

harupy commented on mlflow/fastai/__init__.py:
```diff
@@ -190,6 +193,10 @@ def save_model(
         data=model_data_subpath,
         code=code_dir_subpath,
     )
+    try:
+        mlflow_model.model_size_bytes = get_total_file_size(path)
+    except Exception as e:
+        _logger.info(f"Failed to get the total size of {path} because of error :{e}")
```
Can we warn in ``get_total_file_size` and returns None if it fails?

wenfeiy-db replied:
I thought about try-catching in `get_total_file_size` v.s. doing it here, and slightly prefer doing it here because
* The logic is more concise, as otherwise we'll need both a try-catching and a `if size is not None` block
* Using the logger for each module gives more information

Let me know if you have other opinions

harupy replied:
I thought about something like this:

```suggestion
    if size := get_total_file_size(path):
        mlflow_model.model_size_bytes = size
        
        
def get_total_file_size(...):
    try:
        ...
        return size
    except:
        _logger.warning("Failed")
        return None
```

wenfeiy-db replied:
ahh right - forgot there's the new `:=` syntax.
So I don't really have a strong opinion between the two implementations... I can change if you prefer this more (or if you're ok with both, I'll just be lazy to keep as it is)

harupy replied:
I'd prefer my approach because it looks cleaner and the warning doesn't seem to have to be fired from the flavor module.

harupy commented on mlflow/utils/file_utils.py:
```diff
@@ -967,3 +969,26 @@ def chdir(path: str) -> None:
         os.chdir(cwd)
+
+
+def get_total_file_size(path: str) -> int:
```
can we also accept `pathlib.Path` object?

wenfeiy-db committed:
setter getter

Signed-off-by: wenfeiy-db <wenfei.yan@databricks.com>

wenfeiy-db committed:
move try catch

Signed-off-by: wenfeiy-db <wenfei.yan@databricks.com>

wenfeiy-db committed:
clean

Signed-off-by: wenfeiy-db <wenfei.yan@databricks.com>

wenfeiy-db committed:
precommit

Signed-off-by: wenfeiy-db <wenfei.yan@databricks.com>

wenfeiy-db commented:
@harupy Addressed your comments!

Except for replacing the `os.walk` with `pathlib.Path.rglob`, I'l still trying to understand the reason. See [that](https://github.com/mlflow/mlflow/pull/10110#discussion_r1373978026) thread above.

harupy reviewed (approved):
LGTM!

BenWilson2 commented:
@wenfeiy-db for the failing pyfunc suite test, we can probably just update it to account for this new field entry by doing something similar to this:

```python
reloaded_model_config = Model.load(os.path.join(model_path, "MLmodel"))

    # Check if the reloaded model has a _model_size_bytes attribute that's not None
    if getattr(reloaded_model_config, "_model_size_bytes", None) is not None:
        model_config._model_size_bytes = reloaded_model_config._model_size_bytes

    assert model_config.__dict__ == reloaded_model_config.__dict__
```

wenfeiy-db committed:
fix load_dict

Signed-off-by: wenfeiy-db <wenfei.yan@databricks.com>
